pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,255 @@
+	// run the command and update the status
+	retryErr, cmdErr := e.runWithRetry(ctx, uid, command, retryCount, retryInterval, timeout)
+	if retryErr != nil {
+		switch {
+		case errors.Is(retryErr, context.Canceled):
+			e.log.Info(FenceAgentContextCanceledMessage)
+			return
+		case wait.Interrupted(retryErr):
+			e.log.Info(FenceAgentContextTimedOutMessage)
+		default:
+			e.log.Error(retryErr, FenceAgentRetryErrorMessage)
+		}
 	}
 
-	var (
-		stdoutBuf bytes.Buffer
-		stderrBuf bytes.Buffer
-	)
-
-	//containerName := pod.Spec.Containers[0].Name
-	containerName := "manager"
-
-	// Build the Post request for SPDY (shell) executor
-	req := e.clientSet.CoreV1().RESTClient().Post().
-		Resource("pods").
-		Name(pod.Name).
-		Namespace(pod.Namespace).
-		SubResource("exec").
-		Param("container", containerName)
-
-	req.VersionedParams(&corev1.PodExecOptions{
-		Container: containerName,
-		Command:   command,
-		Stdin:     false,
-		Stdout:    true,
-		Stderr:    true,
-		TTY:       false,
-	}, scheme.ParameterCodec)
-
-	execSPDY, err := remotecommand.NewSPDYExecutor(e.config, "POST", req.URL())
-	if err != nil {
-		e.log.Error(err, "failed building SPDY (shell) executor")
-		return "", "", err
+	if err := e.updateStatusWithRetry(ctx, uid, cmdErr); err != nil {
+		switch {
+		case wait.Interrupted(err):
+			e.log.Info("status context timed out")
+		default:
+			e.log.Error(err, "status retry error")
+		}
 	}
+}
 
-	// Execute the Post request for SPDY (shell) executor
-	err = execSPDY.Stream(remotecommand.StreamOptions{
-		Stdout: &stdoutBuf,
-		Stderr: &stderrBuf,
-		Tty:    false,
-	})
-	if err != nil {
-		e.log.Error(err, "Failed to run exec command", "stdout", stdoutBuf.String(), "stderr", stderrBuf.String())
+func (e *Executer) runWithRetry(ctx context.Context, uid types.UID, command []string, retryCount int, retryInterval, timeout time.Duration) (retryErr, faErr error) {
+	// Run the command with an exponential backoff retry to handle the following cases:
+	// - the command fails: the command is retried until the retryCount is reached
+	// - the command times out: the command is retried until the retryCount is reached
+	// - the FA context times out: the command is cancelled and the status is updated
+	// - the FA context is cancelled: the command is cancelled and the status is not updated
+	// - the command succeeds: the command is not retried and the status is updated
+
+	backoff := wait.Backoff{
+		Steps:    retryCount,
+		Duration: retryInterval,
+		Factor:   1.0,
+	}
+
+	e.log.Info("fence agent start", "uid", uid, "fence_agent", command[0], "retryCount", retryCount, "retryInterval", retryInterval, "timeout", timeout)
+
+	var stdout, stderr string
+	retryErr = wait.ExponentialBackoffWithContext(ctx,
+		backoff,
+		func(ctx context.Context) (bool, error) {
+			ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
```

Comment:
do we need this? :thinking: IIUC, the context we get here is the same which we pass to ExponentialBackoffWithContext. Why would we need to cancel that one when we leave this function. Even more, isn't that  an issue when do that?

Reply:
but maybe it's just too late for me to understand it completely

Reply:
> Why would we need to cancel that one when we leave this function

Not when we leave the function, but when we are in the middle of the function (either in the retry or during a command call), and we want to stop it (e.g. NHC time out)

Reply:
> Not when we leave the function

but that's we do with the defer one line below, not? 🤔